stan/math/prim/mat/vectorize/apply_scalar_binary.hpp

```diff
@@ -0,0 +1,239 @@
+     * template.
+     *
+     * <p>The base class applies to any Eigen dense matrix expression
+     * template.  Specializations define applications to scalars
+     * (primitive or autodiff in the corresponding autodiff library
+     * directories) or to standard library vectors of vectorizable
+     * types (primitives, Eigen dense matrix expressions, or further
+     * standard vectors).
+     *
+     * <p>Each specialization must define the typedef
+     * <code>return_t</code> for the vectorized return type and the
+     * function <code>apply</code> which defines the vectorization or
+     * base application of the function defined statically by the
+     * class F.  The function definition class F defines a static
+     * function <code>fun()</code>, which defines the function's
+     * behavior on scalars.
+     *
+     * @tparam F Type of function to apply.
+     * @tparam T1 Type of first argument to which function is applied.
+     * @tparam T2 Type of second argument to which function is applied.
+     */
+    template <typename F, typename T1, typename T2>
+    struct apply_scalar_binary {
+      /**
+       * Type of underlying scalar for the matrix type T1.
+       * Type of underlying scalar for the matrix type T2.
+       */
+      typedef typename Eigen::internal::traits<T1>::Scalar scalar_t1;
+      typedef typename Eigen::internal::traits<T2>::Scalar scalar_t2;
+
+      /**
+       * Return type for applying the function elementwise to a matrix
+       * expression template of type T1 and T2.
+       */
+      typedef Eigen::Matrix<typename apply_scalar_binary<F, scalar_t1,
+      scalar_t2>::return_t, T1::RowsAtCompileTime, T1::ColsAtCompileTime>
+      return_t;
+
+      /**
+       * Return the result of applying the function defined by the
+       * template parameter F to the specified matrix argument.
+       *
+       * @param x Matrix to which operation is applied.
+       * @param y Matrix to which operation is applied.
+       * @return Componentwise application of the function specified
+       * by F to the specified matrix.
+       */
+      static inline return_t apply(const T1& x, const T2& y) {
+        check_matching_dims<scalar_t1, scalar_t2,
+        T1::RowsAtCompileTime, T1::ColsAtCompileTime,
+        T2::RowsAtCompileTime, T2::ColsAtCompileTime>(
+        "binary vectorization", "x", x, "y", y);
+        return_t result(x.rows(), x.cols());
+        for (int j = 0; j < x.cols(); ++j)
+          for (int i = 0; i < x.rows(); ++i)
+            result(i, j) = apply_scalar_binary<F, scalar_t1, scalar_t2>::
+                           apply(x(i, j), y(i, j));
+        return result;
+      }
+    };
+
+    /**
+     * Template specialization for vectorized functions applying to
+     * int and double arguments.
+     *
+     * @tparam F Type of function defining static apply function.
+     */
+    template <typename F>
+    struct apply_scalar_binary<F, int, double> {
```

Comment:
Is the `int` and `double` here to promote int to double explicitly?  Or is it to avoid ambiguity that might arise without it?  I think a lot of the latter can be solved using `enable_if` on `is_arithmetic` tests (don't know exactly what those look like now in C++11, but we should start using those instead of Boost's).

The thing that's awkward about coding to `int` and `double` is that it doesn't generalize.  I'm lobbying for upgrading all of Stan's internals from `int` to `long`, which means this would all have to be changed if `int` is hardcoded.  If there's any way to avoid hardcoding `int`, it would be a huge help going forward.